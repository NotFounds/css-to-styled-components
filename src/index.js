var fs = require("fs");
var postcss = require("postcss");
var { capitalize, kebabToCamel, snakeToCamel } = require("./utils/cases");

const filePath = process.argv[2];

if (!filePath) {
  console.error(`Please give a css file name.`);
  process.exit(1);
}

if (!fs.existsSync(filePath)) {
  console.error(`${filePath} does not exists`);
  process.exit(1);
}

const scss = fs.readFileSync(filePath, "utf8").replace(/\s*\/\/.*/g, "");

const ast = postcss.parse(scss);

const sources = ast.source.input.css.split("\n");

const result = ast.nodes.map(node => {
  if (node.type === "atrule" && node.name === "import") {
    return;
  }
  const name = node.selector ? node.selector
    .replace(/^\./g, "")
    .replace(/-([a-z])/g, (_, p) => p.toUpperCase()) : node.params;

  const start = node.source.start.line;
  const end = node.source.end.line;
  const defs = sources.slice(start, end - 1);

  let styledFnType = `styled.div\``;
  if (node.type === "atrule" && node.name === "keyframes") {
    styledFnType = `keyframes\``;
  } else if (node.type === "decl" && node.name === undefined) {
    return ["const", node.prop.replace(/^\$/, ""), "=", `"${node.value}";`].join(" ");
  }

  return [
    `const ${capitalize(name)} = ${styledFnType}`,
    ...defs,
    `\``,
  ].join("\n");
});

console.log(
  result
    .join("\n")
    .replace(/:.*\s+\$(.*)\;/g, (_, p) => {
      const cameled = p.replace(/-([a-z])/g, (_, p) => p.toUpperCase());
      return `: \${${cameled}};`;
    })
    .replace(/composes:\s*([a-zA-Z0-9-]+)(\sfrom.*)?/g, (match, p) => {
      return [
        `// ${match}`,
        "  ${" + snakeToCamel(kebabToCamel(p)) + "}",
      ].join("\n");
    })
);
