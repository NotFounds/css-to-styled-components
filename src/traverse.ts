import { ChildNode } from "postcss";
import { capitalize } from "./utils/cases";

export const traverse = (node: ChildNode) => {
  switch (node.type) {
    case "atrule": {
      return;
    }
    case "decl": {
      if (node.prop.match(/^\$[a-zA-Z0-9]+/)) {
        return `const ${node.prop.replace("$", "")} = "${node.value}";`;
      }
      return `  ${node.prop}: ${node.value};`;
    }
    case "rule": {
      const declarations: (string | undefined)[] = node.nodes.map(traverse)
      let [element, name] = node.selector.split(/\.|#/);
      console.log({ element, name });
      const styledFnType = `styled.${element || "div"}`;
      return [
        `const ${capitalize(name)} = ${styledFnType}`,
        ...declarations,
        `\`;`,
      ].join("\n");
    }
    case "comment": {
      return;
    }
  }
};
