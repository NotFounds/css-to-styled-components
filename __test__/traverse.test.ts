import { Declaration, Rule } from "postcss";
import { traverse } from "../src/traverse";

describe("traverse", () => {
  describe("can transform css decl to styled-components", () => {
    it ("transform variable decl", () => {
      const node = new Declaration({ prop: "$color", value: "black" });
      expect(traverse(node)).toMatchSnapshot();
    });
  });
  describe("can transform css rule to styled-components", () => {
    it ("with no element", () => {
      const node = new Rule({
        selector: ".Klass",
        nodes: [
          new Declaration({ prop: "background-color", value: "rgba(0, 0, 0, 0.86)" }),
          new Declaration({ prop: "color", value: "rgb(255, 255, 255)" }),
        ]
      });
      const result = traverse(node);
      expect(result).toMatchSnapshot();
    });

    it ("with element", () => {
      const node = new Rule({
        selector: "p.message",
        nodes: [
          new Declaration({ prop: "background-color", value: "rgba(0, 0, 0, 0.86)" }),
          new Declaration({ prop: "color", value: "rgb(255, 255, 255)" }),
        ]
      });
      const result = traverse(node);
      expect(result).toMatchSnapshot();
    });
  });
});
