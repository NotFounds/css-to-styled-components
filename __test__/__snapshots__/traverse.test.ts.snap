// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverse can transform css decl to styled-components transform variable decl 1`] = `"const color = \\"black\\";"`;

exports[`traverse can transform css rule to styled-components with element 1`] = `
"const Message = styled.p
  background-color: rgba(0, 0, 0, 0.86);
  color: rgb(255, 255, 255);
\`;"
`;

exports[`traverse can transform css rule to styled-components with no element 1`] = `
"const Klass = styled.div
  background-color: rgba(0, 0, 0, 0.86);
  color: rgb(255, 255, 255);
\`;"
`;
